### PR TITLE
Disable autograd in inference-only call paths

### PR DESCRIPTION
**Pull request by gabrieldernbach**
:robot: gabriel-agent

> [!NOTE]
> Depends on #39, which depends on #38. Until those PRs land, GitHub's upstream diff includes the lower layers. Review this incremental change in [`6ac3680...2c9af92`](https://github.com/gabrieldernbach/retrochimera/compare/6ac3680a2396692c54f5d0f82896d93573f2d58f...2c9af92272a96aad902034f94bbfb2a062e4e322). After the lower layers merge, this branch will be rebased onto current `main` so the PR contains only this layer.

## Summary
- Wrap parallel ensemble-worker model calls in `torch.no_grad()`.
- Wrap delegated transformer probability computation in `torch.inference_mode()`.

## Scope
- Incremental layer: 1 commit, 3 files, +63/-2 relative to #39.
- No other inference, training, or model-loading paths are changed.

## Validation
- Added focused coverage verifying gradients are disabled in both changed call paths.
- `python3 -m pytest -q tests/test_beam_state_mapping.py tests/test_run_search.py` — 10 passed.
- `pre-commit run --all-files` — passed.

## Benchmark
The paired end-to-end search benchmark was effectively unchanged: median core time moved from `169.923 s` to `169.744 s` (**0.11%**, below run-to-run variation), while peak allocated and reserved GPU memory fell by about `331 MiB` and `326 MiB`.

The stage-isolation supplement found the intended benefit in `compute_probs`: median latency fell from `2.076 s` to `1.769 s` (**14.77% faster, 1.1733x**), peak allocated memory fell by `5.945 GiB`, and peak reserved memory fell by `10.844 GiB`. Paired outputs were exact.

See the [full protocol, dispersion, and environment details](https://github.com/microsoft/retrochimera/pull/40#issuecomment-5638553274).

### CODE REVIEW

**Comment by gabrieldernbach** [comment]
:robot: gabriel-agent

## Paired GPU benchmark

Benchmarked immediate parent `6ac3680a2396` against PR head `2c9af92272a9` on September 11, 2026.

### End-to-end search

| Metric | Parent | PR head | Change |
|---|---:|---:|---:|
| Core search wall time, median | 169.923 s | 169.744 s | -0.179 s (0.11% faster, 1.0011x) |
| Wrapper wall time, median | 178.779 s | 178.515 s | -0.264 s |
| Peak allocated memory, median | 3.706 GiB | 3.383 GiB | **-330.6 MiB** |
| Peak reserved memory, median | 5.512 GiB | 5.193 GiB | **-326.0 MiB** |

- The timing effect is below run-to-run variation: the head was faster in `3/6` pairs, with a median paired saving of `0.113 s`.
- Normalized outputs were identical across all 12 runs (`4f251f9c554d…`): `24/32` solved targets and `116` model calls.

### Stage isolation

| Stage | Parent median | PR head median | Change |
|---|---:|---:|---:|
| Ensemble `model(...)` | 24.993 s | 24.922 s | 0.28% faster (1.0029x) |
| `compute_probs(...)` | 2.076 s | 1.769 s | **14.77% faster (1.1733x)** |

- Stage A timing is below noise (`4/6` pairs faster), while peak reserved memory fell by **924 MiB**.
- Stage B was faster in **6/6** pairs; median paired saving was **0.306 s** and the two-sided exact sign-test value was `0.03125`.
- Stage B peak allocated memory fell from `8.789` to `2.844 GiB` (**-5.945 GiB**); reserved memory fell from `14.111` to `3.268 GiB` (**-10.844 GiB**).
- Each pair used identical fixed inputs for `compute_probs`; prediction, input, and probability digests matched for every base/head pair.

### Protocol

- Primary: one warmup per side, then six measurements per side in `BAABBAABABBA` order on 32 fixed targets, top-50, at most 200 iterations and 4 model calls per target.
- Stage isolation: six paired 32-target runs producing 1,600 fixed reaction inputs per pair; pair-local payloads prevent unrelated repetitions from invalidating a valid pair.
- Environment: NVIDIA RTX A6000, CUDA device 2, PyTorch `2.2.2+cu121`, Syntheseus `0.8.0`, RDKit `2023.09.6`.

**Conclusion:** this is not a measurable end-to-end search speedup. It is a low-risk inference cleanup with a clear `compute_probs` benefit: about 14.8% lower latency and substantially lower peak GPU memory, with exact paired outputs.

File: tests/test_run_search.py [hunk 1]
```diff
@@ -1,10 +1,70 @@
...
+        def compute_probs(self, reaction_smiles, minibatch_size):
+            assert not torch.is_grad_enabled()
+            return [0.5], [0.5]
+
+    wrapper: Any = object.__new__(AbstractSmilesTransformerModel)  # type: ignore[type-abstract]
+    wrapper.model = Model()
+
+    assert wrapper.compute_probs(["C>>C"], minibatch_size=1) == ([0.5], [0.5])
+
+
```
Comment (lgeiger) [inline tests/test_run_search.py:67]: I don't think these tests are useful, they test implementation details. Let's remove them.

**Review by lgeiger — CHANGES_REQUESTED** [review]
Great! That's a nice free speedup.

Let's also add a changelog entry here

**Comment by gabrieldernbach** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: tests/test_run_search.py (hunk 1 above)
Comment (gabrieldernbach) [reply tests/test_run_search.py]: removed

**Comment by gabrieldernbach** [comment]
added a changelog entry
